src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibraryHelper.java

```diff
@@ -1367,6 +1368,72 @@ private CppCompilationContext initializeCppCompilationContext(CppModel model) {
       }
     }
 
+    if (ruleContext.getFragment(CppConfiguration.class).experimentalEnableImplicitHeaderMaps()) {
+      ImmutableList.Builder<Artifact> headerMapsBuilder = ImmutableList.builder();
+      String targetName = ruleContext.getTarget().getName();
+
+      HeaderMapInfo.Builder internalHeaderMapInfo = new HeaderMapInfo.Builder();
+      internalHeaderMapInfo.addHeaders(publicHeaders.getHeaders());
+      internalHeaderMapInfo.addHeaders(privateHeaders);
+      internalHeaderMapInfo.addHeaders(publicTextualHeaders);
+      Artifact internalHeaderMap = ruleContext.getPackageRelativeArtifact(PathFragment.create(targetName + "_internal.hmap"),
+                ruleContext
+                .getConfiguration()
+                .getGenfilesDirectory(ruleContext.getRule().getRepository()));
+      ruleContext.registerAction(
+        new HeaderMapAction(ruleContext.getActionOwner(),
+            internalHeaderMapInfo.build().getSources(),
+            internalHeaderMap));
+      contextBuilder.addQuoteIncludeDir(internalHeaderMap.getExecPath());
+      headerMapsBuilder.add(internalHeaderMap);
+
+      String namespace;
+      if (ruleContext.attributes().has("header_namespace")) {
```

Comment:
Could we use include_prefix and strip_include_prefix for this? So we don't need another attribute.

Reply:
And for "this" I mean header_namespace attribute.

Reply:
I could take a stab at making it work with the existing `include_prefix` and `strip_include_prefix` attributes in `cc_library` too.

None the less, I think the behavior of `header_namespace` is useful for `objc_library` and especially `objc` that was written under Xcode's conventions.  

The current implementation of `objc_library` -> `header_namespace` has quite different abilities than `cc_library` -> `include_prefix` / `cc_library` -> `strip_include_prefix`. Basically, `header_namespace` puts every single header under a flat namespace regardless of that header's location on disk.

Even if `objc_library` gained the `include_prefix` and `strip_include_prefix` attributes, I don't think it'd be trivial to setup includes for the following repo:

```
LibA
    BUILD
    A1
        SRC1.c
LibB
    BUILD
    C1
        HDR1.h
    C2
        HDR2.h
        D1
            HDR3.h
    .. continues for tens of levels of headers

```

File `LibA/BUILD`

```
objc_library(
    name = "LibA",
    srcs = ["A1/SRC1.c"],
    deps = ["//LibB:LibB"],
)
```

File `LibA/A1/SRC1.c`
```
// The header file resides at LibB/C1/D1/HDR3.h
#import "B/HDR3.h"
```

File `LibB/BUILD`
```
objc_library(
    name = "LibB"
    hdrs = ["C1/HDR1.h", "C2/HDR2.h", "C2/D1/HDR3.h"],

    # Every single header is available under namespace "B"
    # i.e. #include "B/HDR3.h"
    header_namespace = "B"
)

```

When we compile `LibA/SRC1.c` we are passed the following module map:

```
    B/HDR1.h -> LibB/C1/HDR1.h
    B/HDR2.h -> LibB/C2/HDR2.h
    B/HDR2.h -> LibB/C2/D1/HDR2.h
    ... more
```

So includes in File `SRC1.c` are valid
```
// The header file resides at LibB/C1/D1/HDR3.h
#include "B/HDR3.h"
```

In `CocoaPods.org` and other iOS development, this include style is widely used 

Given the usage of includes in iOS development ( i.e. in [Texture](https://github.com/TextureGroup/Texture/blob/master/Source/Details/ASDataController.h) ) and Xcode's include behavior I think it makes sense for  `objc_library`  to have the `header_namespace` attribute.

Reply:
I think I argued for something like header_namespace back in the day; I do think you bring good arguments why it provides an easier user experience. Given that, it might make sense to also support this for C++, as well as supporting it both with and without header maps.

Reply:
Thanks for the feedback @r4nt! For me, in addition to perf benefits, this feature key to making Xcode based source compatible with Bazel in a way that's not to much effort too.

If this is can be upstreamed, then I should be able to add `header_namespace` to C++ library as well.

Ideally, getting `header_namespace` to work without header maps could be broken out into a separate PR. I'm not exactly sure what that would entail, but it seems doable.

Does that sound like a good plan @mhlopko?

Reply:
Is the "flattening" behavior of `header_namespace` really something people rely on? It sounds confusing and error prone to me because of header name collisions. In order to implement this without header map (because people might want to use header_namespace even when they don't want header maps) we would have to pass `-Idir` for every `dir` in the target. I need more convincing that this is good idea @jerrymarino @r4nt :) If we remove "flattening" requirement, then header_namespace behaves exactly like include_prefix. Is there any way to collect more data on how often people rely of the "flattening"?

Reply:
If we want to go into the direction of allowing OS code to work together without changes, it's fundamentally similar to the include_prefix and strip_include_prefix, while being simpler to understand and providing more flexibility, so I'd say it's a clear win.